protocol/meson.build

```diff
@@ -1,103 +1,147 @@
+		'client': true,
+	},
+	'wlr-output-management-unstable-v1': {
+		'path': 'wlr-output-management-unstable-v1.xml',
+	},
+	'wlr-screencopy-unstable-v1': {
+		'path': 'wlr-screencopy-unstable-v1.xml',
+		'client': true,
+	},
+}
 
-client_protocols = [
-	[wl_protocol_dir, 'stable/presentation-time/presentation-time.xml'],
-	[wl_protocol_dir, 'stable/xdg-shell/xdg-shell.xml'],
-	[wl_protocol_dir, 'unstable/idle-inhibit/idle-inhibit-unstable-v1.xml'],
-	[wl_protocol_dir, 'unstable/linux-dmabuf/linux-dmabuf-unstable-v1.xml'],
-	[wl_protocol_dir, 'unstable/pointer-constraints/pointer-constraints-unstable-v1.xml'],
-	[wl_protocol_dir, 'unstable/pointer-gestures/pointer-gestures-unstable-v1.xml'],
-	[wl_protocol_dir, 'unstable/relative-pointer/relative-pointer-unstable-v1.xml'],
-	[wl_protocol_dir, 'unstable/tablet/tablet-unstable-v2.xml'],
-	[wl_protocol_dir, 'unstable/text-input/text-input-unstable-v3.xml'],
-	[wl_protocol_dir, 'unstable/xdg-decoration/xdg-decoration-unstable-v1.xml'],
-	[wl_protocol_dir, 'unstable/xdg-shell/xdg-shell-unstable-v6.xml'],
-	'idle.xml',
-	'input-method-unstable-v2.xml',
-	'wlr-export-dmabuf-unstable-v1.xml',
-	'wlr-foreign-toplevel-management-unstable-v1.xml',
-	'wlr-gamma-control-unstable-v1.xml',
-	'wlr-input-inhibitor-unstable-v1.xml',
-	'wlr-layer-shell-unstable-v1.xml',
-	'wlr-screencopy-unstable-v1.xml',
-]
+foreach name, proto : protocols
+	path = proto.get('path')
 
-wl_protos_src = []
-wl_protos_headers = []
-
-foreach p : protocols
-	xml = join_paths(p)
-	wl_protos_src += custom_target(
-		xml.underscorify() + '_server_c',
-		input: xml,
+	code = custom_target(
+		path.underscorify() + '_c',
+		input: path,
 		output: '@BASENAME@-protocol.c',
 		command: [wayland_scanner, 'private-code', '@INPUT@', '@OUTPUT@'],
 	)
-	wl_protos_headers += custom_target(
-		xml.underscorify() + '_server_h',
-		input: xml,
+	wlr_files += code
+
+	wlr_files += custom_target(
+		path.underscorify() + '_server_h',
+		input: path,
 		output: '@BASENAME@-protocol.h',
 		command: [wayland_scanner, 'server-header', '@INPUT@', '@OUTPUT@'],
 	)
-endforeach
 
-foreach p : client_protocols
-	xml = join_paths(p)
-	wl_protos_headers += custom_target(
-		xml.underscorify() + '_client_h',
-		input: xml,
-		output: '@BASENAME@-client-protocol.h',
-		command: [wayland_scanner, 'client-header', '@INPUT@', '@OUTPUT@'],
-	)
-endforeach
-
-lib_wl_protos = static_library(
-	'wl_protos',
-	wl_protos_src + wl_protos_headers,
-	dependencies: wayland_client.partial_dependency(compile_args: true),
-)
+	if proto.get('client', false)
+		header = custom_target(
+			path.underscorify() + '_client_h',
+			input: path,
+			output: '@BASENAME@-client-protocol.h',
+			command: [wayland_scanner, 'client-header', '@INPUT@', '@OUTPUT@'],
+		)
 
-wlr_protos = declare_dependency(
-	link_with: lib_wl_protos,
-	sources: wl_protos_headers,
-)
+		set_variable(name.underscorify() + '_c', code)
+		set_variable(name.underscorify() + '_h', header)
```

Comment:
I'm not a very big fan of `set_variable` overall, because it makes it difficult to understand where they come from. I'd prefer to have a dict instead (e.g. `protocols_code.get(name)` would return the file for the generated code for instance).

Reply:
>I'd prefer to have a dict instead (e.g. protocols_code.get(name) would return the file for the generated code for instance).

That's actually a pretty good idea. Unfortunately meson (0.52.1) can't use a non string literal for a dictionary key:
```python
protocol_code = {}
foreach proto, path : protocols
    # ...
    protocol_code += { proto: code } # This is invalid
endforeach
```
And due to the way dictionaries work (i.e. they're pure) and some dictionary-of-dictionaries type stuff, I can't really think of a way to work around this (e.g. putting it in the protocols dictionary).

But this will be fixed in 0.53.0: https://mesonbuild.com/Syntax.html#dictionaries
It has been left as set_variable/get_variable currently, but changed to use an array of protocol strings in the wayland backend.
I think this should be revisited when the next meson version is out and we eventually target it for whatever reason.

Reply:
Makes sense. Created a new issue so that we don't forget about it.